src/tools/non-promotion-reasons.md

```diff
@@ -1,45 +1,380 @@
+
+**The cause:**
+You're trying to promote a variable that might have been
+written to since it was promoted.
+
+Example:
+
+{:.bad}
+{% prettify dart tag=pre+code %}
+f(bool b, int? i, int? j) {
+  if (i != null) return;
+  if (b) {
+    i = j;           // (1)
+  }
+  if (!b) {
+    print(i.isEven); // (2) ERROR
+  }
+}
+{% endprettify %}
+
+In this example, when flow analysis hits (1),
+it demotes `i` from non-nullable `int` back to nullable `int?`.
+A human can tell that the access at (2) is safe
+because there's no code path that includes both (1) and (2), but
+flow analysis isn't smart enough to see that,
+because it doesn't track correlations between
+conditions in separate `if` statements.
+
+You might fix the problem by combining the two `if` statements:
+
+{:.good}
+{% prettify dart tag=pre+code %}
+f(bool b, int? i, int? j) {
+  if (i != null) return;
+  if (b) {
+    i = j;
+  } else {
+    print(i.isEven);
+  }
+}
+{% endprettify %}
+
+Note that in straight-line control flow cases like these (no loops),
+flow analysis takes into account the right hand side of the assignment
+in order to decide whether to demote.
+That means you could also fix this by changing the type of `j` to `int`.
+
+
+### Possibly written in a previous loop iteration {#loop-or-switch}
+
+**The cause:**
+You're trying to promote something that
+might have been written to in a previous iteration of a loop,
+and so the promotion was invalidated.
+
+Example:
+
+{:.bad}
+{% prettify dart tag=pre+code %}
+f(Link? p) {
+  if (p != null) return;
+  while (true) {    // (1)
+    print(p.value); // (2) ERROR
+    var next = p.next;
+    if (next == null) break;
+    p = next;       // (3)
+  }
+}
+{% endprettify %}
+
+When flow analysis reaches (1),
+it looks ahead and sees the write to `p` at (3).
+But because it's looking ahead,
+it hasn't yet figured out the type of the right-hand side of the assignment,
+so it doesn't know whether it's safe to retain the promotion.
+To be safe, it invalidates the promotion.
+
+You might fix this problem by moving the null check to the top of the loop:
+
+{:.good}
+{% prettify dart tag=pre+code %}
+f(Link? p) {
+  while (p != null) {
+    print(p.value);
+    p = p.next;
+  }
+}
+{% endprettify %}
+
+This situation can also arise in `switch` statements if
+there's a `case` block that has a label,
+because you can use labeled `switch` statements to construct loops:
+
+{:.bad}
+{% prettify dart tag=pre+code %}
+f(int i, int? j, int? k) {
+  if (j == null) return;
+  switch (i) {
+    label:
+    case 0:
+      print(j.isEven);
+      j = k;
+      continue label;
+  }
+}
+{% endprettify %}
+
+
+### In catch after possible write in try
+
+**The cause:**
+The variable might have been written to in a `try` block,
+and execution is now in a `catch` block.
+
+Example:
+
+{:.bad}
+{% prettify dart tag=pre+code %}
+f(int? i, int? j) {
+  if (i == null) return;
+  try {
+    i = j;                 // (1)
+    if (i == null) return; // (2)
```

Comment:
Consider adding some placeholder text in the try block so that it looks more plausible that an exception might actually occur, e.g.:

```dart
try {
  i = j; // (1)
  ... other logic ...
  if (i == null) return; // (2)
  ... other logic ...
} catch (...) {
  print(i.isEven); // ERROR
}
```